posthog/api/person.py

```diff
@@ -391,6 +394,70 @@ def destroy(self, request: request.Request, pk=None, **kwargs):
         except Person.DoesNotExist:
             raise NotFound(detail="Person not found.")
 
+    @extend_schema(
+        parameters=[
+            OpenApiParameter(
+                "delete_events",
+                OpenApiTypes.BOOL,
+                description="If true, a task to delete all events associated with this person will be created and queued. The task does not run immediately and instead is batched together and at 5AM UTC every Sunday (controlled by environment variable CLEAR_CLICKHOUSE_REMOVED_DATA_SCHEDULE_CRON)",
+                default=False,
+            ),
+            OpenApiParameter(
+                "distinct_ids",
+                OpenApiTypes.OBJECT,
+                description="A list of distinct ids. We'll delete all persons associated with those distinct ids. The maximum amount of ids you can pass in one call is 100.",
+            ),
+            OpenApiParameter(
+                "ids",
+                OpenApiTypes.OBJECT,
+                description="A list of PostHog person ids. We'll automatically fetch all related persons and delete those. The maximum amount of ids you can pass in one call is 100.",
+            ),
+        ],
+    )
+    @action(methods=["POST"], detail=False, required_scopes=["person:write"])
+    def bulk_delete(self, request: request.Request, pk=None, **kwargs):
+        """
+        This endpoint allows you to bulk delete persons, either by the PostHog persons ID or by distinct IDs. You can pass through a maximum of 100 ids per call.
+        """
+        if request.data.get("distinct_ids"):
+            if len(request.data["distinct_ids"]) > 100:
+                raise ValidationError("You can only pass 100 distinct_ids in one call")
+            persons = self.get_queryset().filter(persondistinctid__distinct_id__in=request.data.get("distinct_ids"))
+        elif request.data.get("ids"):
+            if len(request.data["ids"]) > 100:
+                raise ValidationError("You can only pass 100 ids in one call")
+            persons = self.get_queryset().filter(uuid__in=request.data["ids"])
+        else:
+            raise ValidationError("You need to specify either distinct_ids or ids")
+
+        for person in persons:
+            delete_person(person=person)
+            self.perform_destroy(person)
```

Comment:
I was worried about deletions failing halfway through, because we can't do transactions in clickhouse. So if we first bulk-deleted all the persons in posthog, but failed to half the people in clickhouse, we'd end up in a weird state. By doing it sequentially, at least the failure will be contained to one person, rather than potentially up to 100.

Reply:
Hm, I don't think we can tell if a ClickHouse-side delete fails anyway, because all `delete_person()` does is queue a deletion row into Kafka, which is extremely unlikely to fail. So if we first bulk-delete in Postgres, and _then_ emit deletion rows to CH, that should be the highest level of integrity possible in this situation.

Reply:
It's definitely a tradeoff – in this explicit bulk delete case it doesn't feel great to put this O(n) load on Postgres, but I see what you mean. For maximum integrity this route, we should swap the `delete_person(person=person)` and `self.perform_destroy(person)` lines though.

Reply:
Hm this doesn't work as the person gets deleted before clickhouse get a chance to delete it. I also think it's more likely for clickhouse to fail, thus the current order does make sense